proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,468 @@
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br TextMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "command":
+				br.command = v
+				br.commands = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+				br.keyExactBytes = []byte(v)
+			case "keyPrefix":
+				br.keyPrefix = v
+				br.keyPrefixBytes = []byte(v)
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				proxylib.ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		if br.command == "" {
+			proxylib.ParseError("command not specified", rule)
+		}
+		log.Infof("Parsed TextMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// TextMemcacheParserFactory implements proxylib.ParserFactory
+type TextMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *TextMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("TextMemcacheParserFactory: Create: %v", connection)
+	return &TextMemcacheParser{connection: connection, replyQueue: make([]*replyIntent, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &TextMemcacheParserFactory{}
+
+var textMemcacheParserFactory *TextMemcacheParserFactory
+
+const (
+	parserName = "textmemcache"
+)
+
+func init() {
+	log.Info("init(): Registering textMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, textMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7TextMemcacheRuleParser)
+}
+
+// TextMemcacheParser implements proxylib.Parser
+type TextMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	replyQueue   []*replyIntent
+}
+
+type replyIntent struct {
+	command   []byte
+	requestID uint32
+	denied    bool
+}
+
+var _ proxylib.Parser = &TextMemcacheParser{}
+
+// OnData parses binary memcached data
+func (p *TextMemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	log.Infof("OnData with offset %d", offset)
+
+	if reply {
+		if p.injectFromQueue() {
+			return proxylib.INJECT, len(DeniedMsg)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	// TODO: don't copy data to new slices
+	data := (bytes.Join(dataBuffers, []byte{}))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	linefeed := bytes.Index(data, []byte("\r\n"))
+	if linefeed < 0 {
+		log.Infof("Did not receive full first line, asking for 1 byte more")
+		return proxylib.MORE, 1
+	}
+
+	// TODO: iterate over data without copying it to new slices
+	// Tokenizing in memcached is done by spaces: https://github.com/memcached/memcached/blob/master/memcached.c#L2978
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := textMemcacheMeta{
+			command: tokens[0],
+		}
+
+		frameLength := 0
+		noreply := false
+		if p.isCommandRetrieval(meta.command) {
```

Comment:
Consider replacing all of these `if/else` with a `switch case/case/case..`

Reply:
The thing is I am not switching on just one value here, but every case is much different.

Reply:
Should be fine for a bare `switch` and then `case`s with the boolean expressions. It is just another way of writing a long list of exclusive if blocks.